DOC Improve `plot_target_encoder_cross_val.py` example (#26677)
scikit-learn/scikit-learn: doc/glossary.rst

        exceptional behaviours on the estimator using semantic :term:`estimator
        tags`.

    cross-fitting
    cross fitting
        A resampling method that iteratively partitions data into complementary
        'train' and 'estimation' subsets. The 'estimation' subset may be used
        to estimate a parameter or predict a target. The outputs from

>         each iteration are combined and used in a downstream prcess, generally

Reviewer: ```suggestion
        each iteration are combined and used in a downstream process, generally
```

>         to avoid bias.

Reviewer 2: Maybe it would be useful to cross-link to typical (meta-)estimators that implement this strategy (e.g. StackingRegressor/Classifier and TargetEncoder, CalibratedClassifierCV...).

Note: the current implementation of `CalibratedClassifierCV` is a bit different from the others because it does not attempt to concatenate the predictions of the first stage to fit the second stage (but instead fits one second stage estimator per cross-fitting iteration and then averages the predictions of the second stage model). Still, I think those are morally all examples of a cross-fitting strategy.

Another example not yet implemented in scikit-learn are honest trees and forests where the values of the leafs are estimated on different training samples that those used to learn the decision splits of the trees.

Author: Good idea, I meant to do this but  forgot!

>         A resampling method that iteratively partitions data into complementary
>         'train' and 'estimation' subsets. The 'estimation' subset may be used
>         to estimate a parameter or predict a target. The outputs from
>         each iteration are combined and used in a downstream prcess, generally
>         to avoid bias.

Reviewer 2: I don't find the "train" and "estimation" names very clear because one could argue that training and estimating are similar. Maybe it would be helpful to speak in terms of first and second stage as follows:

```suggestion
        A resampling method that iteratively partitions data into mutually exclusive
        subsets to fit a two stage estimator. The second stage is fit on predictions
        of the first stage computed on a subset of data not seen when training the
        first stage. The objective is to avoid having any overfitting in the first
        stage introduce a bias in the training data distribution of the second stage.
```

Author: 'mutually exclusive' is much clearer and more explicit than 'complementary', have used this in the cross validation entry too, thanks!
Is `TargetEncoder` technically a 2 stage estimator? There isn't really a second stage, it just outputs the transformed X values for use by another estimator...? I have thus tried to make it more general and re-worded. Not entirely happy though and open to changes. Also I may have mis-understood some aspects!

Reviewer 2: `TargetEncoder` is the first stage and the HistGradientBoosting model (or anything else) you train on the output of `TargetEncoder` is the second stage.

For stacking and calibration with use meta-estimator, whereas for `TargetEncoder` we use a transformer pipeline and implement the cross-fitting only in the `fit_transform` of the first stage transformer.

Author: Thanks, yes that is my understanding too.

Now I realise my mis-understanding is over what 'two stage estimator' means. Initially I took it to mean that a single estimator that performs two stages, which none of the estimators do (`CalibratedClassifierCV` technically takes a `estimator` argument which does the first stage). Whereas you meant 'two stage estimator' as an estimator that is designed to perform a 2 stage process. What about changing the wording to be something like: 
"to fit a two stage estimator chain" or "to fit an estimator in a two stage process/model" ??

Author: ping @ogrisel , would either suggestion be suitable?

    cross-validation
    cross validation
        A resampling method that iteratively partitions data into complementary
        'train' and 'test' subsets so model performance can be evaluated on
        unseen data. This conserves data as avoids the need to hold out a
        'validation' dataset and accounts for variability as multiple rounds of
        cross validation are genreally performed.

>         See :ref:`User Guide <_cross_validation>` for more details.

Reviewer: The CI is failing due to the first underscore.

```suggestion
        See the :ref:`User Guide <cross_validation>` for more details.
```

>         See :ref:`User Guide <_cross_validation>` for more details.

Reviewer 2: There is a problem with this reference:

```
The following documentation warnings may have been generated by PR #https://github.com/scikit-learn/scikit-learn/pull/26677:
/home/circleci/project/doc/glossary.rst:200: WARNING: undefined label: '_cross_validation'
```

I believe there should be no leading underscore.

Author: Yes you are right, thanks. Bad copy paste

>         See :ref:`User Guide <_cross_validation>` for more details.

Reviewer 2: ```suggestion
        See the :ref:`user guide <cross_validation>` for more details.
```
